### ISSUE TITLE
Flexible font define in style

### ISSUE THREAD

**Issue by alanchenboy** [maplibre/maplibre-style-spec#868]
# Design Proposal: Flexible font define in style

## Motivation

MapLibre inherits the font pbf design from Mapbox, this is a limit to rendering complex text. We must let the core reach the font file to finish shaping and rendering complex text like Indic, Khmer Burmese, etc.

## Proposed Change

The design looks like CSS unicode-range, 

  ```
 "font-faces": [
        {
            "text-font": ["Noto Sans Regular"],
            "font-files": [
                {
                    "url": "https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/khmer.ttf",
                    "unicode-range": [
                        "U+1780-17FF"
                    ]
                },
                {
                    "url":"https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/myanmar.ttf",
                    "unicode-range": [
                        "U+1000-109F", "U+A9E0-0xA9FF", "U+AA60-0xAA7F"
                    ]
                },
                {
                    "url":"https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/devanagari.ttf",
                    "unicode-range": [
                        "U+0900-097F", "U+A8E0-0xA8FF"
                    ]
                }
                
            ]
        }
    ],
```
full style:
https://alanchenboy.github.io/harfbuzzResource/hindi.json

## API Modifications

A new feature for complex font rendering, it's transparent for the map users.  

Here is an example of how to render complex text on the native side.  
https://github.com/maplibre/maplibre-native/pull/1439.  
## Migration Plan and Compatibility

N/A

## Rejected Alternatives

N/A

**Comment by HarelM** [comment]
Thanks for taking the time to open this design proposal.
Can you clarify the relation between this property and `glyphs`? It is not clear when one should be selected over the other.
Also, will using this require some configuration in the rendering part? special build? a plugin?
We don't have other style properties that has external requirements in order to work as far as I know...

**Comment by louwers** [comment]
We would use Harfbuzz for text shaping in MapLibre Native by default, so no special plugins would be required.

**Comment by HarelM** [comment]
Wouldn't this require a special build? Or is the default build comes with Harfbuzz and you can reduce build size if you need, but this would be the special build - i.e without it?

**Comment by louwers** [comment]
No, we would enable it by default, although it would be possible to drop out if you have binary size constraints and you make your own builds.

**Comment by alanchenboy** [comment]
This approach means that if a font is defined here by font-face, then the selected Unicode point will be shaped with Harfbuzz and then rasterized with FreeType.
In addition to rendering complex fonts, this also has the implicit benefit of saving traffic by the font files actually smaller than the pre-rendered pbf files.

**Comment by wipfli** [comment]
Exactly! Additionally, if we have an API to define server-generated SDF glyphs we can benefit from:
- Clients MapLibre GL JS and MapLibre Native do not have to use FreeType
- Bezier-based SDF generation which is more precision and looks better than TinySDF-based SDF generation

The API could maybe look something like this:

```json
"font-faces": [
      {
          "text-font": ["Noto Sans Regular"],
          "font-files": [
              {
                  "url": "https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/khmer.ttf",
                  "unicode-range": [
                      "U+1780-17FF"
                  ],
                  "glyphs": "https://alanchenboy.github.io/harfbuzzResource/glyphs//Noto%20Sans%20Regular/khmer/{range}.pbf"
              },
              {
                  "url":"https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/myanmar.ttf",
                  "unicode-range": [
                      "U+1000-109F", "U+A9E0-0xA9FF", "U+AA60-0xAA7F"
                  ],
                  "glyphs": "https://alanchenboy.github.io/harfbuzzResource/glyphs//Noto%20Sans%20Regular/myanmar/{range}.pbf"
              },
              {
                  "url":"https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/devanagari.ttf",
                  "unicode-range": [
                      "U+0900-097F", "U+A8E0-0xA8FF"
                  ],
                  "glyphs": "https://alanchenboy.github.io/harfbuzzResource/glyphs//Noto%20Sans%20Regular/devanagari/{range}.pbf"
              }
              
          ]
      }
  ],
```

Some fonts have several hundred glyphs so it might make sense to split the glyphs up into ranges of 256 like we do with normal MapLibre fonts. The difference to normal fonts would be that the index is actually the glyph index from the font .ttf file and not simply the unicode codepoint.

**Comment by 1ec5** [comment]
> Bezier-based SDF generation which is more precision and looks better than TinySDF-based SDF generation

FYI, maplibre/maplibre-gl-js#4564 would lean more heavily on TinySDF, and I have a half-baked idea in https://github.com/maplibre/maplibre-gl-js/pull/4564#discussion_r1730061958 to use it exclusively if the top-level `glyphs` property is missing (which is currently an error). The idea would be to rely on Web Fonts (or the native equivalent) for any custom fonts, without requiring any MapLibre-specific server-side changes.

**Comment by HarelM** [comment]
How would this proposal work for native?
I know @1ec5 is working on a branch to support this in web, but I want to understand how can this use native fonts from the OS if possible for native?
I think the range part is important to define, but I'm good with most of the solutions above, I prefer the array with two items in it (start and end as separate array entries), also it should probably be called `unicode-ranges` (plural).
Besides that, aligning to CSS definitions is a good idea as it will create clarity from people familiar with this area and would like to apply the same idea here.

**Comment by 1ec5** [comment]
> The only major concern I remember during the Eastern call was ensuring we don't break existing styles.

I believe this was in response to the approach I started with https://github.com/maplibre/maplibre-gl-js/pull/4564#discussion_r1730061958 and continued in 1ec5/maplibre-gl-js#1 relies on relaxing an existing hard requirement in the style specification. Graceful degradation isn’t the only way to achieve the goal of maintaining compatibility with existing stylesheets.

In order to adopt extended TinySDF, a style author would intentionally break compatibility with existing versions of MapLibre on all platforms by removing the heretofore required `glyphs` property from the stylesheet. It would be akin to incrementing the `version`, except that they wouldn’t have to do that because well-formed clients would already recognize upfront whether they’re compatible with the stylesheet.

If we really want, we could implement a fallback strategy on platforms that don’t support extended TinySDF to fall back to a last resort fontstack, either bundled or distributed from maplibre.org. But I don’t think this is necessary, because omitting `glyphs` will be a very intentional step on the part of the style author. Doing so can come with caveats.

The alternative would be that style authors have to continue specifying glyph PBFs that they have to generate but hope will never be used. This will become a much larger burden on developers once we expand character support beyond the Basic Multilingual Plane. I don’t think developers will see this as a good tradeoff.

> how can this use native fonts from the OS if possible for native?

This is infeasible on the Web and unlikely to ever be feasible. Browser vendors are trying to lock down access to native fonts to prevent fingerprinting. The experimental [Local Font Access API](https://developer.mozilla.org/en-US/docs/Web/API/Local_Font_Access_API) is only supported by Chromium and requires user permissions. I don’t think MapLibre should require the user to give privacy-related permissions just to view a map. More likely, we’d continue to improve upon the TinySDF technique on the Web to the point that it’s closer to true native text rendering on the native platforms using Harfbuzz.

**Comment by HarelM** [comment]
I was asking about using OS fonts on native devices using the maplibre-native implantation and how would this spec proposal cater for it. Downloading fonts is less likely for native apps as far as I know (although not my expertise).

**Comment by 1ec5** [comment]
I think it’s much more feasible on native platforms (though you may need to perform expensive [reconstructive surgery](https://stackoverflow.com/a/9381326)). It would be inherently native-only, permanently, as opposed to the situation with extended TinySDF, which would only be a temporary situation so that we wouldn’t have to synchronize releases across all three platforms.

I suspect direct system font access is kind of redundant on iOS anyways, since you can already get the vectors of the glyph if you really want. And both Android and iOS now provide direct native access to text shaping.

**Comment by 1ec5** [comment]
The other aspect of direct font access is licensing. Mapbox’s original idea behind glyph PBFs is that they’re just collections of screenshots, which under U.S. law are ineligible for copyright protection. As soon as you depend on actual font data, the font’s license kicks in. Fortunately, these days, there’s a wealth of fonts licensed for distribution as Web fonts (think Google Fonts), while the landscape for “desktop” fonts is more restrictive. Presumably system fonts are fair game, because you don’t have to embed anything, but then you’re at the mercy of the device’s font selection, which can be highly variable even on iOS, let alone Android.

**Comment by HarelM** [comment]
I'm not sure I'm following to be honest. 
If the defined spec change in this thread is only relevant for web and can't be applied to native, we should consider adding options to map construction in web and not change the spec.
If this is applicable to native, how would this work in theory, can native use tinysdf as well? Does this mean native will download the file and apply it to the map?

**Comment by louwers** [comment]
@HarelM I agree, the spec should be cross-platform and should not have web-specific or native-specific extensions. 

> Would this would work with local URLs (native only, I assume?). I could see wanting to ship a font file or glyphs locally on mobile.

MapLibre Native is able handle `asset://` and `file://` URLs. Although as @1ec5 says default system fonts are inaccessible on iOS. You would need to bundle a font with your app (or download it to the filesystem).

> If this is applicable to native, how would this work in theory, can Native use tinysdf as well? Does this mean native will download the file and apply it to the map?

Native is already using TinySDF for CJK characters. But @alanchenboy PR implements rasterization using FreeType.

---

So this may be a stupid suggestion, I don't know much about the complexities of font rendering. But can we push the complexity to the font server and update the `glyphs` template?

```json
{
  "glyphs": "https://maplibre.org/font/{fontstack}/{range}.{font_format}?unicode_block={unicode_block}"
}
```

For the `{font_format}` MapLibre GL JS can substitute `.pbf` to request server-generated SDFs (or ignore the `glyphs` property and use TinySDF when configured with some map property), MapLibre Native can request a `.ttf`.

And `{unicode_block}` would be substituted with the [block name](https://unicode.org/Public/UCD/latest/ucd/Blocks.txt) e.g. `Basic_Latin`, `Khmer`, `Myanmar_Extended-B`, etc.

**Comment by 1ec5** [comment]
> Although as @1ec5 says default system fonts are inaccessible on iOS. You would need to bundle a font with your app (or download it to the filesystem).

Right, I think this would be a limitation of anything that needs to access the actual font data. By default, on Android and iOS, the TinySDF-based [local font rendering](https://maplibre.org/maplibre-native/ios/latest/documentation/maplibre-native-for-ios/customizing_fonts#Client-side-fonts) mechanism for CJK only needs rasterized characters and metrics, which the standard library already provides. Extended TinySDF wouldn’t change much about this; the main difference is just how we’d segment the string to send to TinySDF.

Incidentally, the existing CJK mechanism relies on a separate configuration file that overrides what’s in the stylesheet, equivalent to the map constructor option suggested by @HarelM. Putting the option in the style specification would’ve posed a barrier to some cross-platform applications: developers generally want to use system-provided CJK fonts rather than bundling huge custom fonts, but each platform comes with a different set of CJK fonts.

**Comment by 1ec5** [comment]
> Graceful degradation isn’t the only way to achieve the goal of maintaining compatibility with existing stylesheets. My approach relies on relaxing an existing “optional” requirement in the style specification.

maplibre/maplibre-style-spec#1045 elaborates on the ideas that I’ve been floating around in various tickets. It doesn’t completely cover the use case for the proposal here, but I hope the alternative approach I’ve laid out would be able to accommodate whatever we might need for native text layout on the native platforms.

**Comment by 1ec5** [comment]
#1045 makes the case that, however important the `font-faces` information is to being able to integrate HarfBuzz, it should be the responsibility of application code using a runtime API. During the technical committee calls yesterday, #1045 was challenged on the grounds that specifying all the subset URLs and codepoint ranges in application code would be quite laborious and error-prone compared to this proposal. I don’t think the two proposals are mutually exclusive, but I have concerns about the practicality of this proposal.

I believe that defining TrueType file URLs at design time optimizes for too specific a workflow and font set. Web developers strongly prefer WOFF over uncompressed TrueType or OpenType. Neither type foundries like MonoType and Adobe nor CDNs like Google Fonts would even provide direct access to a font file at a fixed URL. Even if we make `font-faces` the first (?) inherently native-only addition to the style specification, I think it confuses the role of the style designer with the frontend developer. A style designer knows what typeface they want for each layer, so they need to be able to define that in the stylesheet rather than application code. By contrast, a frontend developer knows where the font resources would live, so they need to be able to define that at build time or runtime rather than expecting it to come predefined in the stylesheet.

If we truly need the font locations to be defined in a JSON file, then I think we’re potentially looking at the prospect of a new JSON format alongside style JSON and TileJSON. This would be analogous to the CSS stylesheets that the likes of MonoType and Google Fonts provide to Web developers. However, I don’t think `font-faces` as proposed will be as broadly useful to mobile developers as supposed. Mobile developers who want to bundle the font with the application will need to specify a different URL depending on the platform, so this proposal would force them to fork the style in two. More size-conscious developers would instead prefer to load the font at runtime as an [on-demand resource](https://developer.apple.com/documentation/xcode/configuring-custom-fonts) (iOS) or [downloadable font](https://developer.android.com/develop/ui/views/text-and-emoji/downloadable-fonts) (Android), but these mechanisms don’t provide access the URL until the font has finished downloading. So regardless, there will be a need for a font definition API at runtime.

**Comment by louwers** [comment]
> Even if we make font-faces the first (?)  inherently native-only addition

Would not be the first. 😉 https://maplibre.org/maplibre-style-spec/sources/#volatile

> I think it confuses the role of the style designer with the frontend developer. A style designer knows what typeface they want for each layer, so they need to be able to define that in the stylesheet rather than application code. By contrast, a frontend developer knows where the font resources would live, so they need to be able to define that at build time or runtime rather than expecting it to come predefined in the stylesheet.

While this may be true in some cases, I think the many frontend developers would actually prefer to have the option to just set a map style (using a URL provided by their tile provider) and not have to worry at all about font distribution at all. Tile providers probably also want to make it as uncomplicated as possible to use their styles. Furthermore, having a `font-faces` property as part of the style makes updating or changing the fonts that a style uses possible.

> this proposal would force them to fork the style in two

Developers that want to concern themselves with font distribution would have the option to extend (or override) `font-faces` with `asset://` or `file://` URLs at runtime with this proposal.

---

I like your approach thinking about personas. Maybe we can get some tile/style providers in here to share their thoughts.


Pinging @hyperknot (OpenFreeMap), @ianthetechie (Stadia Maps), @hy9be (could you pass this thread to the relevant people at AWS Location Services?):

tl;dr of this thread: The future of font-loading for MapLibre GL JS seems clear: use the web platform (#1045). For MapLibre Native we also want to move offer an alternative font loading mechanism (via TrueType fonts and shaping with Harfbuzz). This will allow your styles to be used in many new parts of the world where they don't use Latin or CJK scripts. The question is: do you want to (have the option) to specify in the style where these TrueType fonts can be downloaded from? Or is font distribution a concern that should be separate from the style?

**Comment by hyperknot** [comment]
For OpenFreeMap I'm planning to implement the following:
1. Offer a set of known open source fonts. For example NotoSans generated from sources in a few variants, hosted next to the tiles, on the same domain, for the ease of use.
2. Adding a pattern-matching wildcard path in nginx, so that every non-existing font requested will arrive at a failsafe location, for example on NotoSans Bold.

By this I mean if the requested path is:
```
https://tiles.openfreemap.org/fonts/_non_existing_font_/65024-65279.pbf
```

Then it'll be served as if it was:
```
https://tiles.openfreemap.org/fonts/Noto%20Sans%20Bold/65024-65279.pbf
```

For the future, it'd be nice to have this fallback mechanism somehow in the specs/client-side, so that it could work on hosts which do not support such wildcard pattern-matching. For example, today, if you put the fonts on S3 or Cloudflare R2 buckets, it is not possible today to do this. The simplest "serverless" architecture which can work today I believe is Cloudflare Workers to do this pattern-matching.

Now, where to put the fallback location is an interesting question. If you put it in the style JSON, then it won't be portable between hosts. I actually think the best place for this would be the TileJSON, as that's the only thing what the tile provider has control over.

So practically speaking, I can imagine a TileJSON like this:

```
{
"tilejson": "3.0.0",
"tiles": ["https://tiles.openfreemap.org/planet/20250305_001002_pt/{z}/{x}/{y}.pbf"],
"fallback_fonts": ["https://tiles.openfreemap.org/fonts/Noto%20Sans%20Bold/{range}.pbf"],

```

**Comment by 1ec5** [comment]
> I think the many frontend developers would actually prefer to have the option to just set a map style (using a URL provided by their tile provider) and not have to worry at all about font distribution at all. Tile providers probably also want to make it as uncomplicated as possible to use their styles. Furthermore, having a `font-faces` property as part of the style makes updating or changing the fonts that a style uses possible.

My impression is that a mobile developer who wants a very turnkey experience would already find system fonts acceptable at the outset, since that’s what they normally work with already. But of course this may not be true of the style designer. I’m sympathetic to the need for style designers to have control over the look and feel of the map, but I don’t really think a style designer would find a large font mapping like this particularly approachable either. The logical conclusion is [hundreds of lines of configuration](https://github.com/osm-americana/fontstack66/blob/78bc33cd4a292513e21ec70fdf177529c1cef906/fonts.json) for a pan-language font family like Noto. On the other hand, one of the benefits of relying on each platform’s native font selection capabilities is that we can get font substitutions for free, without any configuration: Noto Sans for Latin and Cyrillic, Hiragino for Japanese, etc.

By the way, until now, many independently hosted MapLibre stylesheets have chosen style-specific fontstack names, pro forma, in contrast to styles hosted by services that provide more generic fontstacks. In theory, this gives them the ability to swap out the actual typeface without literally changing anything in the stylesheet, but that probably creates some fun for caching, and I’m unsure how much of a practical need there is to avoid touching the stylesheet. I’m expecting this practice to wane over time as `text-font` starts to be interpreted as PostScript font names. (And surely a customer of a hosted service would be unhappy if “Noto” started referring to not-Noto overnight.)

> Developers that want to concern themselves with font distribution would have the option to extend (or override) `font-faces` with `asset://` or `file://` URLs at runtime with this proposal.

OK, I wasn’t seeing any reference to a runtime API in this proposal or maplibre/maplibre-native#1439.

> The future of font-loading for MapLibre GL JS seems clear: use the web platform (#1045). For MapLibre Native we also want to move offer an alternative font loading mechanism (via TrueType fonts and shaping with Harfbuzz). This will allow your styles to be used in many new parts of the world where they don't use Latin or CJK scripts.

To be clear, #1045 is a first step to expanding the TinySDF codepath to _all_ of Unicode on _all_ platforms as the baseline text rendering option, with support for complex text shaping. Although #1045 links to my Web-based proof of concept in 1ec5/maplibre-gl-js#1, I don’t foresee any technical hurdles for implementing the same functionality on Android and iOS.[^mapbox] The existing glyph PBF mechanism would remain for backwards compatibility and for developers who care about pixel-for-pixel cross-platform compatibility and higher rendering fidelity. Nevertheless, I see TinySDF on the native platforms as merely a stopgap while we wait for HarfBuzz (a la maplibre/maplibre-native#1439) across all writing systems. Ultimately, the TinySDF mechanism is a Rube Goldberg machine, but it can represent a significant improvement over the status quo.

> Adding a pattern-matching wildcard path in nginx, so that every non-existing font requested will arrive at a failsafe location, for example on NotoSans Bold.

#1045 would make this unnecessary, since the client would fall back to a system font.

> I actually think the best place for this would be the TileJSON, as that's the only thing what the tile provider has control over.

Besides TileJSON, there’s also the sprite JSON. Would there be a technical reason not to spin off a separate file for this font mapping?

[^mapbox]: And it looks like Mapbox already has, though it relies on an override in the API that probably hamstrings its usage.

**Comment by louwers** [comment]
@hyperknot Thanks for pitching in. The question is as follows: in the future we will likely allow loading fonts in other ways besides hosted `.pbf` files. On MapLibre GL JS you will have the option to use web fonts which are converted in a format so they can be rendered on the fly (with TinySDF). With MapLibre Native you will be able to use TrueType fonts. Now we are at a fork in the road, because do we want to continue specifying where fonts can be found in the `style.json`? On the web we can rely on fonts that are available in the environment, on mobile this is not always possible, so this PR suggests a new property to specify were TrueType fonts can be downloaded from.

> Besides TileJSON, there’s also the sprite JSON. Would there be a technical reason not to spin off a separate file for this font mapping?

That's an interesting idea. This way people could host a single endpoint with a whole bunch of available fonts with vetted licenses. Although defining and implementing a new standard is much more involved. And we would probably still have to add a new property to the style.

**Comment by hyperknot** [comment]
@louwers thank for explaining.

I can imagine something along the lines of:
- Specifying client-dependent font stack in style JSON, like in CSS.
```
"my_special.ttf", Avenir, Montserrat, Corbel, 'URW Gothic', source-sans-pro, sans-serif;
```
https://modernfontstacks.com/ is a great resource for cross-platform usage

- If CSS font stacks are not compatible with TinySDF, then I guess you need to specify the exact TrueType / WOFF2 fonts in some of the recommended formats in this thread (abs URL, rel URL, local file, data)

- Add the failback mechanism in TileJSON. I see no reason to introduce a new JSON file for this, I think we are talking about 1-2 values here, definitely not something which would fill a JSON file.

**Comment by 1ec5** [comment]
> On the web we can rely on fonts that are available in the environment, on mobile this is not always possible

This is a limitation of maplibre/maplibre-native#1439 due to needing access to the font’s raw vector and hinting data, but this is not an inherent limitation to font access overall. MapLibre Native can and already does access fonts available in the environment, and extended TinySDF would take advantage of that fact for more writing systems.

It is true that a specific font available on one platform may not be available on another. This is why the TinySDF mechanism on every platform already interprets `text-font` as a fallback list. Just as when designing a webpage, a style author would list the superset of desired fonts from each platform, possibly ordered by aesthetics or language coverage.

> Add the failback mechanism in TileJSON. I see no reason to introduce a new JSON file for this, I think we are talking about 1-2 values here, definitely not something which would fill a JSON file.

The proposal here is for a file that would specify a different font file per writing system. In some cases, such as with the Noto fonts, this file would stretch to 500–600 lines before minification.

TileJSON shouldn’t be tightly coupled to fonts. A given tileset can be used in a variety of styles with different aesthetics. Conversely, a given style may pull in multiple tilesets. I would’ve suggested extending the sprite JSON format to avoid proliferating file formats, but unfortunately that format isn’t extensible.

**Comment by williamscraigm** [comment]
Note, Arial Unicode MS only covers the Unicode BMP (Basic Multilingual Plane) - it's incomplete even as a fallback

**Comment by prusswan** [comment]
My main needs are Unicode support, and for font display to "just work" like regular web applications (reasonable fallback to system fonts even when no fonts/glyhs are specified or available. A single font (or even a few fonts) to cover the entire space is certainly desirable but I expect most users would not need it.

**Comment by louwers** [comment]
```
"font-faces": {
  "Noto Sans Regular": [
    {
      "url": "https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/khmer.ttf",
      "unicode-ranges": [
        [
          "1780",
          "17FF"
        ]
      ]
    },
    {
      "url": "https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/myanmar.ttf",
      "unicode-ranges": [
        [
          4096,
          "109F"
        ],
        [
          "A9E0",
          "A9FF"
        ],
        [
          "AA60",
          "AA7F"
        ]
      ]
    },
    {
      "url": "https://alanchenboy.github.io/harfbuzzResource/fonts/Noto%20Sans%20Regular/devanagari.ttf",
      "unicode-ranges": [
        [
          "0900"
        ],
        [
          "097F"
        ],
        [
          "A8E0"
        ],
        [
          "A8FF"
        ]
      ]
    }
  ]
}
```
